README.md
# demo.todomvc

Demo project to implement [TodoMVC](http://todomvc.com/) using Bract/Gossamer.


## Usage

### Setup the database

You must setup the database as a one-time task before running the application:

```shell
$ lein do clean, run -m liquibase update
```

The `project.clj` file has a shorter alias defined for this task, so you may run the following command instead:

```shell
$ lein liquibase update
```


### Starting the server

Once you start the server you should visit http://localhost:3000 in your web browser.

You can build the app locally and start the server as follows:

```shell
$ lein do clean, uberjar         # implicitly calls `lein cljsbuild once`
$ java -jar target/uberjar/demo.todomvc-0.6.2-0.3.0-standalone.jar -vf config/config.edn
```

Alternatively, you may want to run it as follows in development (but logging configuration may be out of control):

```shell
$ lein do clean, cljsbuild auto  # in a separate terminal
$ lein ring server-headless      # ensure not to include clean task
```

Should you need a browser REPL during development, run the following commands instead (with logging config caveat):

```shell
$ lein do clean, figwheel        # in a separate terminal, preferably with rlwrap
$ lein ring server-headless      # ensure not to include clean task
```

### Triggering batch job

The TODO items are only soft-deleted in the database by the application. The CLI command `purge` hard-deletes such
deleted entries, which can be triggered as follows:

```shell
$ java -jar target/uberjar/demo.todomvc-0.6.2-0.3.0-standalone.jar -c purge -vf config/config.edn
```


### REPL based development

Be sure to start the ClojureScript code compilation using `lein do clean, cljsbuild auto` or `lein do clean, figwheel`
ahead of starting the REPL with `lein repl`.


Run `lein do clean, repl` to launch the REPL (run `(start)` to start the server):

```clojure
bract.dev.repl=> (help)  ; shows the help screen
```

### Inspect and toggle logs

Running the application should by default emit logs in the `logs` folder. To enable/disable certain kinds of logs
you may override the default configuration by specifying as such in the config file:

```clojure
"logback.appender.console.enabled"  "true"
"logback.appender.jsonfile.enabled" "true"
"logback.appender.textfile.enabled" "false"
```

**Note:** Jetty server logs (lein-ring) are configured in `test/clj/jetty-logging.properties` to appear in console only.


## License

Copyright © 2017-2021 [Shantanu Kumar](https://github.com/kumarshantanu)

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# demo.todomvc Change Log

## Todo & Ideas

- [Todo] Drop Liquibase
- [Idea] Use PostgreSQL
- [Idea] Use Reagent
- [Todo] Use bi-directional routes


## 0.6.2-0.3.0 / 2021-February-28

- Bract
  - Use Bract 0.6.2
  - Process CLI args in DEV mode just as main mode
  - Use `bract.dev.repl` for REPL support
    - Use port 3001 for nREPL
    - Drop `dev` files
  - Use `bract.core.dev` as main entrypoint in DEV
- Application
  - Replace discontinued `goog.dom.query` with `js/document.querySelectorAll`
  - Move `dev/liquibase.clj` to `test/clj/liquibase.clj` (uptake Bract 0.6.2)
  - Use separate URIs for `/todos/*` and `/todo/:id/*` (conflict resolution for Calfpath 0.8.0)
- Version upgrades
  - Clojure 1.10.2
  - ClojureScript 1.10.758
  - ring-core 1.8.2
  - ring-jetty-adapter (commented) 1.8.2
  - http-kit 2.5.1
  - cljs-ajax 0.8.1
  - cljstache 2.0.6
  - H2 database 1.4.200
  - asphalt 0.6.7


## 0.6.1-0.2.0 / 2019-January-10

- Upgrade Clojure to version 1.10.0
- Use Gossamer 0.6.1-0.2.0
  - Trim down bract-context.edn
  - Trim down bract-context.dev.edn
  - Remove `bract.cli`, `bract.ring` modules
  - Remove redundant `app.version` key definition
  - Remove redundant print-version command


## 0.6.1 / 2018-October-10

- Use Bract 0.6.1
- Use Gossamer 0.6.1
- Upgrade CLJS to 1.10.339
- Bump Aleph version to 0.4.6
- Fix disabling JS minification in dev
- Use global atom to hold runtime state


## 0.6.0 / 2018-May-18

- Use Gossamer 0.6.0
  - Use logback-included in logback.xml
  - Use bract.ring and Gossamer config resources
  - Move all inducer reference from config files to context files
  - Use Calfpath routes instead of direct dispatch
- Replace Clostache (unmaintained) with Cljstache
  - https://github.com/fotoetienne/cljstache
- Add shutdown hook and global uncaught exception handler


## 0.5.1 / 2018-March-05

- Use Bract 0.5.1
- Add fallback config (in `baseconfig.edn`) for application version and hostname
- Use `bract.cli.main` as CLI entry point
- Use soft-delete to remove TODO items in database
- Add CLI command to purge soft-deleted TODO items
- Use `bract.core.dev-init` to initialize test namespaces


## 0.5.0 / 2018-February-18

- Update dependencies
  - Bract 0.5.0
  - Clojure 1.9.0, ClojureScript 1.9.946
  - Asphalt 0.6.5
  - Calfpath 0.5.0
  - Cambium: API 0.9.1 and Logback module 0.4.1
  - Ring 1.6.3
  - Cljs-ajax 0.7.3
- Use bract.ring wrappers
  - JSON output for /info and /health endpoints
  - Display app version in /info result


## 0.4.1 / 2017-August-08

- Use Bract 0.4.1
- Bump Asphalt version to 0.6.3
- Use new Asphalt 0.6.2 option `:result-set-worker` in tests


## 0.4.0 / 2017-August-05

- Use Bract 0.4.0
- Bump CLJS version to 1.9.854
- Bump Asphalt version to 0.6.2
- Move global/stateful vars from `demo.todomvc.config` to `demo.todomvc.global` namespace
- Rename namespace `demo.todomvc.config` to `demo.todomvc.keydef`
- Implement `version` CLI command using custom `demo.todomvc.main/-main` entrypoint
- Make logging config work when using _lein-ring_
  - Workaround for issue https://github.com/weavejester/lein-ring/issues/190
- Improve log archive management by capping the total archive size


## 0.3.1 / 2017-June-30

- Use Clostache template on the server-side to render HTML with minified JS
- Update ClojureScript version to 1.9.660
- Replace implicit page load with JavaScript `window.onload` event handler
  - Ref: https://github.com/Day8/re-frame/blob/master/examples/todomvc/resources/public/index.html
  - Ref: https://github.com/Day8/re-frame/blob/master/examples/todomvc/project.clj
- Use Hiccups CLJS library (instead of string concatenation) to generate TODO item HTML
- Hide unnecessary controls when there are no TODO items
- Implement toggle-complete-all feature
- Use Bract 0.3.1 components


## 0.3.0 / 2017-June-19

- ClojureScript based browser DOM manipulation
- File based H2 database storage
- Server-side web endpoints to talk to the database
- Database migration using Liquibase
- File based application and metrics logging
- Use Bract 0.3.0 components


project.clj
(defproject bract/demo.todomvc "0.6.2-0.3.0"
  :description "Demo TodoMVC app using Clojure, ClojureScript and the Bract/Gossamer framework"
  :url "https://github.com/bract/demo.todomvc"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :resource-paths ["resources" "target/generated/resources"]  ; see :project-edn entry
  :source-paths ["src/clj"]
  :test-paths   ["test/clj"]
  :pedantic?    :warn
  :dependencies [[org.clojure/clojure "1.10.2"]
                 ;; ----- ClojureScript -----
                 [org.clojure/clojurescript "1.10.773" :exclusions [com.google.code.findbugs/jsr305]]
                 [cljs-ajax "0.8.0" :exclusions [com.fasterxml.jackson.dataformat/jackson-dataformat-smile
                                                 com.fasterxml.jackson.core/jackson-core
                                                 cheshire]]  ; for making AJAX calls from within the browser
                 [hiccups   "0.3.0"]  ; for dynamically generating HTML
                 ;; ----- server-side web -----
                 [bract/gossamer.core "0.6.2-0.3.0"]
                 [ring/ring-core      "1.8.2" :exclusions [commons-codec]]
                 [cljstache           "2.0.6"]  ; mustache templates
                 ;; ----- web servers (uncomment any one) -----
                 ;;[aleph                   "0.4.6" :exclusions [org.clojure/tools.logging]]
                 [http-kit                "2.5.1"]
                 ;;[org.immutant/immutant   "2.1.10"]
                 ;;[ring/ring-jetty-adapter "1.8.2"]
                 ;; ------ database -----
                 [com.h2database/h2 "1.4.200"]  ; the embedded H2 database
                 [cumulus  "0.1.2"]  ; for easily deriving JDBC connection params
                 [clj-dbcp "0.9.0"]  ; for making database connection pool
                 [asphalt  "0.6.7"]  ; for reading/writing databases using JDBC
                 ]
  :target-path "target/%s"
  :plugins [[lein-cljsbuild   "1.1.8" :exclusions [org.clojure/clojure]]
            [lein-figwheel    "0.5.20"]
            [lein-project-edn "0.2.0"]
            [lein-ring        "0.12.5"]]
  :hooks [leiningen.project-edn/activate]
  :clean-targets ^{:protect false} ["resources/public/js/out"
                                    "resources/public/js/app.js"
                                    "resources/public/js/app.min.js"
                                    :target-path]
  :project-edn {:output-file "target/generated/resources/project.edn"
                :verify-edn? true}
  :ring {:handler bract.ring.dev/handler
         :init    bract.ring.dev/init!
         :port    3000
         :nrepl   {:start? true :port 3001}}
  :profiles {:dev     {:dependencies [[bract/bract.dev "0.6.2-0.2.0" :exclusions [org.clojure/tools.reader]]
                                      [clj-liquibase   "0.6.0"]]
                       :main ^:skip-aot bract.core.dev
                       :repl-options {:init-ns bract.dev.repl
                                      :port 3001}
                       :cljsbuild {:builds [{:id "dev"
                                             :source-paths ["src/cljs"]
                                             :figwheel {:on-jsload "demo.todomvc.app/main"}
                                             :compiler {:main demo.todomvc.app
                                                        :asset-path "/public/js/out"
                                                        :output-to  "resources/public/js/app.js"
                                                        :output-dir "resources/public/js/out"
                                                        :pretty-print  true
                                                        :optimizations :none
                                                        :source-map    true
                                                        :source-map-timestamp true}}]}}
             :uberjar {:aot [bract.core.main]
                       :main ^:skip-aot bract.core.main
                       :hooks [leiningen.cljsbuild]
                       :cljsbuild {:builds [{:id "prod"
                                             :source-paths ["src/cljs"]
                                             :compiler {:output-to  "resources/public/js/app.min.js"
                                                        :output-dir "target/js/out"
                                                        :pretty-print  false
                                                        :optimizations :advanced}}]}
                       :pedantic? :abort}}
  :aliases {"liquibase"  ["run" "-m" "liquibase"]})

